### PR TITLE
Service Fabric: update to use SiloHostBuilder

### PR DESCRIPTION
**Pull request by ReubenBond**
Brings a much needed update to Service Fabric support, primarily to support `SiloHostBuilder` and `ClientBuilder` and separate hosting from membership.

* Splits the `Microsoft.Orleans.ServiceFabric` into `Microsoft.Orleans.Hosting.ServiceFabric` and `Microsoft.Orleans.Membership.ServiceFabric`. Membership depends on hosting.
* The `Microsoft.Orleans.ServiceFabric` package will no longer be created.

Updated samples are in a separate branch, will open PR when we bump the version.
Example usage:
```C#
protected override IEnumerable<ServiceInstanceListener> CreateServiceInstanceListeners()
{
    // Listeners can be opened and closed multiple times over the lifetime of a service instance.
    // A new Orleans silo will be both created and initialized each time the listener is opened and will be shutdown 
    // when the listener is closed.
    var listener = OrleansServiceListener.CreateStateless(
        (serviceContext, builder) =>
        {
            // Use Service Fabric for cluster membership.
            builder.AddServiceFabricMembership(serviceContext);

            // Optional: configure logging.
            builder.ConfigureLogging(logging => logging.AddDebug());

            var config = new ClusterConfiguration();
            config.Globals.RegisterBootstrapProvider<BootstrapProvider>("poke_grains");
            config.Globals.ReminderServiceType = GlobalConfiguration.ReminderServiceProviderType.ReminderTableGrain;

            // Service Fabric manages port allocations, so update the configuration using those ports.
            config.Defaults.ConfigureServiceFabricSiloEndpoints(serviceContext);

            // Tell Orleans to use this configuration.
            builder.UseConfiguration(config);

            // Add your application assemblies.
            builder.AddApplicationPart(typeof(ICalculatorGrain).Assembly);

            // Alternative: add all loadable assemblies in the current base path (see AppDomain.BaseDirectory).
            builder.AddApplicationPartsFromBasePath();
        });

    return new[] { listener };
}
```
xref: #1059

### CODE REVIEW

File: src/Orleans.Hosting.ServiceFabric/ConfigurationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+using System.Collections.ObjectModel;
+using System.Fabric;
+using System.Fabric.Description;
+using System.Net;
+using Orleans.ServiceFabric;
+using NodeConfiguration = Orleans.Runtime.Configuration.NodeConfiguration;
+
+namespace Orleans.Hosting
+{
+    public static class ConfigurationExtensions
```
Comment (jdom) [inline src/Orleans.Hosting.ServiceFabric/ConfigurationExtensions.cs]: Please use a class name that has `ServiceFabric` in it. Especially since this goes to the default Orleans.Hosting namespace where collisions are likely to occur

Comment (ReubenBond) [reply src/Orleans.Hosting.ServiceFabric/ConfigurationExtensions.cs]: good call, will fix

File: src/Orleans.Hosting.ServiceFabric/FabricSiloInfo.cs [hunk 2]
```diff
@@ -11,24 +12,6 @@
     public class FabricSiloInfo
```
Comment (jdom) [inline src/Orleans.Hosting.ServiceFabric/FabricSiloInfo.cs]: I see that this Json-serializable class is also use with IOptions. Should they really be the same thing? Haven't reviewed the whole thing to understand what's the JSON representation used for

Comment (ReubenBond) [reply src/Orleans.Hosting.ServiceFabric/FabricSiloInfo.cs]: The JSON representation is used for Service Fabric's naming service.

The IOptions portion is mostly so that external developers could publish their own endpoints via the `OtherEndpoints` property.

File: src/Orleans.Hosting.ServiceFabric/OrleansCommunicationListener.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        public OrleansCommunicationListener(Action<ISiloHostBuilder> configure)
+        {
+            this.configure = configure ?? throw new ArgumentNullException(nameof(configure));
+        }
+
+        /// <summary>
+        /// Gets or sets the underlying <see cref="ISiloHost"/>.
+        /// </summary>
+        /// <remarks>Only valid after <see cref="OpenAsync"/> has been invoked. Exposed for testability.</remarks>
+        public ISiloHost SiloHost { get; private set; }
```
Comment (jdom) [inline src/Orleans.Hosting.ServiceFabric/OrleansCommunicationListener.cs]: I would just call this property Host, so that we can hopefully then switch the type to use the generic IHost

File: src/Orleans.Hosting.ServiceFabric/FabricSiloInfo.cs (hunk 2 above)
Comment (ReubenBond) [reply src/Orleans.Hosting.ServiceFabric/FabricSiloInfo.cs]: BTW, originally I thought of having two, but I figured it was unnecessary/over-engineering

Comment (jdom) [reply src/Orleans.Hosting.ServiceFabric/FabricSiloInfo.cs]: I'm not really convinced, but don't feel that strongly either, especially since we haven't full migrated off the legacy configuration and into Options. I think we can re-asses afterwards

Comment (ReubenBond) [reply src/Orleans.Hosting.ServiceFabric/FabricSiloInfo.cs]: What's your preference for now? Change the PR or leave it?

Comment (jdom) [reply src/Orleans.Hosting.ServiceFabric/FabricSiloInfo.cs]: nah, leaving it for now it's OK. I was just waiting for the build to finish

File: src/Orleans.Membership.ServiceFabric/Orleans.Membership.ServiceFabric.csproj [hunk 4]
```diff
@@ -1,15 +1,15 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <PackageId>Microsoft.Orleans.ServiceFabric</PackageId>
```
Comment (jdom) [inline src/Orleans.Membership.ServiceFabric/Orleans.Membership.ServiceFabric.csproj]: We should probably create the `Microsoft.Orleans.ServiceFabric` meta-package. Do we do it now or in a different PR? I suppose it's better to do it in the same PR where we split the package

Comment (ReubenBond) [reply src/Orleans.Membership.ServiceFabric/Orleans.Membership.ServiceFabric.csproj]: Oh originally I had the metapackage, but I figured it was just noise. I can add it again.

Comment (jdom) [reply src/Orleans.Membership.ServiceFabric/Orleans.Membership.ServiceFabric.csproj]: Yeah, I think we should do that

**Comment by jdom** [comment]
@dotnet-bot test functional
